### PR TITLE
ENG-550: Slack - support Socket Mode

### PR DESCRIPTION
**Pull request by daabr**

### CODE REVIEW

File: integrations/slack/websockets/socketmode.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+	api := slack.New(botToken, slack.OptionAppLevelToken(appToken))
+	webSocketClients[appID] = socketmode.New(api)
+
+	go func() {
+		for {
+			smh := socketmode.NewSocketmodeHandler(webSocketClients[appID])
+			smh.HandleDefault(h.socketModeHandler)
+			err := smh.RunEventLoop()
+			if err != nil {
+				h.logger.Error("Slack Socket Mode (re)connection error", zap.Error(err))
```
Comment (itayd) [inline integrations/slack/websockets/socketmode.go]: ?

Comment (daabr) [reply integrations/slack/websockets/socketmode.go]: Bug fixed - `continue`, not `return` if `RunEventLoop()` returns (i.e. reconnection error).

Also reversed condition to use an implicit `continue`, as requested.

Verified - goroutine is indeed needed.

Not done: can't merge `err := ...` and `if` now that the usage of `err` is outside the condition block

File: internal/backend/logger/logger.go [hunk 2]
```diff
@@ -15,7 +15,7 @@ type Config struct {
 	Zap zap.Config `koanf:"zap"`
 }
 
-var defaultZapLevel = zap.NewAtomicLevelAt(zap.InfoLevel)
```
Comment (itayd) [inline internal/backend/logger/logger.go]: different pr

File: integrations/slack/websockets/socketmode.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+		logger:        l,
+		secrets:       sec,
+		dispatcher:    d,
+		scope:         scope,
+		integrationID: id,
+	}
+}
+
+// Key = Slack app ID (to ensure one WebSocket per app).
+var webSocketClients = make(map[string]*socketmode.Client)
```
Comment (itayd) [inline integrations/slack/websockets/socketmode.go]: mutex

File: integrations/slack/websockets/socketmode.go [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+			zap.Any("request", e.Request),
+		)
+	}
+}
+
+func (h handler) dispatchAsyncEventsToConnections(tokens []string, event *eventsv1.Event) {
+	ctx := extrazap.AttachLoggerToContext(h.logger, context.Background())
+	for _, connToken := range tokens {
+		event.IntegrationToken = connToken
+		event := kittehs.Must1(sdktypes.EventFromProto(event))
```
Comment (itayd) [inline integrations/slack/websockets/socketmode.go]: comment why must

File: integrations/slack/server.go [hunk 5]
```diff
@@ -1,48 +1,49 @@
...
+	tokens, err := s.List(context.Background(), "slack", "websockets")
+	if err != nil {
+		l.Error("Failed to list WebSocket tokens", zap.Error(err))
+		return
+	}
+
+	for _, connToken := range tokens {
+		data, err := s.Get(context.Background(), "slack", connToken)
+		if err != nil {
+			l.Error("Missing data for Slack Socket Mode app", zap.Error(err))
```
Comment (itayd) [inline integrations/slack/server.go:44]: return

File: integrations/slack/websockets/bot_events.go [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+// See https://api.slack.com/apis/connections/events-api#responding.
+// Compare this function with the [webhooks.HandleBotEvent] implementation.
+func (h handler) handleBotEvent(e *socketmode.Event, c *socketmode.Client) {
+	defer c.Ack(*e.Request)
+
+	// Reuse the Slack event's JSON payload instead of the struct.
+	body, err := e.Request.Payload.MarshalJSON()
+	if err != nil {
+		h.logger.Error("Bad request from Slack websocket",
+			zap.Any("payload", e.Request.Payload),
```
Comment (itayd) [inline integrations/slack/websockets/bot_events.go:27]: return

Comment (daabr) [reply integrations/slack/websockets/bot_events.go:27]: done also in `interactive.go`

**Review by itayd — APPROVED** [review]

**Pull request merged**